File: leetcode/0350-intersection-of-two-arrays-ii/solution.py
from collections import Counter
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        counts = Counter(nums1)
        result = []

        for x in nums2:
            if counts[x] > 0:
                result.append(x)
                counts[x] -= 1

        return result

File: leetcode/0350-intersection-of-two-arrays-ii/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size()) {
            swap(nums1, nums2);
        }

        unordered_map<int, int> counts;
        for (int x : nums1) {
            counts[x]++;
        }

        vector<int> result;
        for (int x : nums2) {
            auto it = counts.find(x);
            if (it != counts.end() && it->second > 0) {
                result.push_back(x);
                it->second--;
            }
        }

        return result;
    }
};
